### PR TITLE
Convert backend tests to use pytest

### PR DESCRIPTION
**Pull request by QuLogic**
Net line count drop, seems good to me.

Also, fix one of the decorators to work properly with pytest.

### CODE REVIEW

File: lib/matplotlib/tests/test_backend_bases.py [hunk 1]
```diff
@@ -81,10 +79,6 @@ def test_get_default_filename_already_exists():
         open(os.path.join(test_dir, 'image.png'), 'w').close()
 
         filename = canvas.get_default_filename()
-        assert_equal(filename, 'image-1.png')
+        assert filename == 'image-1.png'
     finally:
         shutil.rmtree(test_dir)
-
-if __name__ == "__main__":
```
Comment (NelleV) [inline lib/matplotlib/tests/test_backend_bases.py]: How do we run tests on only one module with pytest?

**Comment by NelleV** [comment]
I don't know pytest well. How do you run tests on only one module with pytest?

**Comment by QuLogic** [comment]
If it's installed from source, you can use the file itself: `pytest lib/matplotlib/tests/test_backend_bases.py` or if you want to test the installed version `pytest --pyargs matplotlib.tests.test_backend_bases`.

**Comment by QuLogic** [comment]
To give these individual PRs some substance, I have not done just a straight conversion, but also tried to take advantage of pytest's additional niceties. I can do a PR of just straight nose-removal if there's interest in that.

**Comment by NelleV** [comment]
I think this underlines that our test suite is very complex, which unfortunately means also complicated to maintain. I also don't find pytest very nice nor readable to use, but it might just be a question of getting used to it.

**Comment by phobson** [comment]
@NelleV it is a very different approach, for sure. I tried for ~18 months to like pytest, and just couldn't get into it.

Then about a year ago I forced myself to start from scratch with it on a new package and it clicked and made python magical again.

**Comment by QuLogic** [comment]
I added one more change to fix the determinism tests from the backends which are not fully included in the `test_backend_*` files.

**Comment by phobson** [comment]
What in tarnation is taking the CI so long on this? I'll kick to the see if that helps.

**Comment by QuLogic** [comment]
You might have missed it, but I rebased because of the conflicts.

**Comment by phobson** [comment]
This looks good to me.

**Comment by phobson** [comment]
Oh, sorry @QuLogic . The Github UI said the commits were from 6 days ago. I can't blame them for that. But it was confusing nonetheless.

**Pull request merged**